teps/0008-trigger-crd.md

```diff
@@ -0,0 +1,271 @@
+
+This is a TEP for the implementation of the Trigger CRD that were
+discussed in the following docs:
+
+- [Trigger Listener CRDs](https://docs.google.com/document/d/1zWpmEhtSNe8KAPKvTJE7Pjg5Uzk8mx9MUjSN24D1jUg/edit#heading=h.emfu2q8r4dx8)
+- [Multitenant EventListener](https://docs.google.com/document/d/1NX0ExhPad6ixTM8AdU0b6Vc3MVD5hQ_vIrOs9dIXq-I/edit)
+
+## Motivation
+
+Today, EventListener commonly are created for every namespace and handle TriggerBinding and
+TriggerTemplate in that same namespace. Creation of EventListener causes pod proliferation. For every
+namespace that requires handling of webhook events like github, we need EventListener which in turns
+lead to this pod proliferation and causes excess resource consumption. In multitenant scenario where
+projects are divided according to namespaces, we require EventListener for each namespaces. So for
+100 of projects, we would need 100 of pods just for acting as sink for these webhook events like github.
+These pods in turn consume resources.
+
+### Goals
+
+1. Reduction in resource cosumption due to EventListener.
+2. Ability to have single EventListener to cater to whole cluster.
+3. Separation of Concerns :  The more ELs there are, the more additional configuration are needed - in some cases this additional bit of configuration is something that a cluster admin/operator would perform not an user/app developer.
+For declarative config as code scenarios, it would be easier for a user to have create the Trigger configuration part as part of config with code. The cluster admin/operator can create the EventListeners, expose it to a public address, etc. And the same listener can be shared among many Triggers. This does reduce the number of listener Pods but also allows for a better separation of concerns among the user and the operator.
+
+### Non-Goals
+
+How to create Multitenant EventListener resource with appropriate permissions won't be part of this proposal.
+This will be handled by operator or admin of the cluster.
+
+## Requirements
+
+1. EventListener can create Tekton resources in a single namespace or across multiple namespaces if configured to do so.
+2. Binding of these cross namespace Trigger to EventListener. When an EL recieves an event, it can process configured subset of triggers using path/selector.
+3. EL can still process triggers defined in the current form aka backwards compatibility.
+
+## Proposal
+
+EventListener CRD will be split into two - a Trigger CRD and a EventListener CRD.
+
+### Trigger CRD
+A single Trigger defines configuration for processing events i.e. it consists of TriggerBindings,
+a single TriggerTemplate, and optionally interceptors and a serviceAccount.
+```
+apiVersion: v1alpha1
+kind: Trigger
+metadata:
+  name: my-repo-trigger
+  labels:
+    eventlistener: operator
+spec:
+  serviceAccountName: "blah"
+  interceptors:
+    cel:
+      filter: "$(header.eventType == "push")"
+  bindings:
```

Comment:
The binding syntax here is the one I proposed in https://github.com/tektoncd/triggers/issues/617

I think we should use the one that we actually support for now 😄

Reply:
Good call. Fixed that.